### PR TITLE
HDDS-9541. NPE in OMDBCheckpointServlet.Lock#lock with OM Ratis disabled

### PR DESCRIPTION
**Pull request by aierate**
## What changes were proposed in this pull request?
When ratis is disabled, a null pointer exception is reported, so I tried to fix it.
I am setting ozone.om.ratis.enable in ozone.xml as :
//ozone-site.xml
```
<property>
      <name>ozone.om.ratis.enable</name>
      <value>false</value>
</property> 
```
When I using http://my.om.address:9874/dbCheckpoint

This is the error:
```
2023-10-25 20:17:09,444 [qtp1013364696-226] ERROR org.apache.hadoop.hdds.utils.DBCheckpointServlet: Unable to process metadata snapshot request. 
java.lang.NullPointerException
        at org.apache.hadoop.ozone.om.OMDBCheckpointServlet$Lock.lock(OMDBCheckpointServlet.java:672)
        at org.apache.hadoop.hdds.utils.DBCheckpointServlet.generateSnapshotCheckpoint(DBCheckpointServlet.java:197)
        at org.apache.hadoop.hdds.utils.DBCheckpointServlet.doGet(DBCheckpointServlet.java:303)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)
        at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1656)
        at org.apache.hadoop.http.lib.StaticUserWebFilter$StaticUserFilter.doFilter(StaticUserWebFilter.java:110)
        at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
        at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1626)
        at org.apache.hadoop.hdds.server.http.HttpServer2$QuotingInputFilter.doFilter(HttpServer2.java:1660)
        at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
        at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1626)
        at org.apache.hadoop.hdds.server.http.NoCacheFilter.doFilter(NoCacheFilter.java:48)
        at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
        at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1626)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:552)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:600)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1440)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:505)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1355)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
        at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.server.Server.handle(Server.java:516)
        at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:487)
        at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:732)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:479)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
        at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
        at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:409)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
        at java.lang.Thread.run(Thread.java:748)
```
I am fixed it.
## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9541

### CODE REVIEW

**Comment by adoroszlai** [comment]
Thanks @aierate for the patch.  Can you please explain how it was tested?  Is there any unit or integration test you could modify to verify the fix?

**Review by hemantk-12 — COMMENTED** [review]
Thanks for the patch @aierate

Was it an HA env where you were disabling ratis by setting `ozone.om.ratis.enable`? If I'm not wrong this code path is only reached in HA env and Ratis should be enabled in that scenario?

**Comment by aierate** [comment]
> Thanks @aierate for the patch. Can you please explain how it was tested? Is there any unit or integration test you could modify to verify the fix?

Thanks @adoroszlai . 
 Here are the results of my integration test when I set 
```
<property>
      <name>ozone.om.ratis.enable</name>
      <value>false</value>
</property> 
```
and perform 
http://my.om.address:9874/dbCheckpoint
```
2023-11-09 12:03:30,409 [qtp70666339-235] INFO org.apache.hadoop.hdds.utils.DBCheckpointServlet: R
eceived GET request to obtain DB checkpoint snapshot
2023-11-09 12:03:30,419 [qtp70666339-235] INFO org.apache.hadoop.ozone.om.OMDBCheckpointServlet: C
ompaction pausing 1 started.
2023-11-09 12:03:30,461 [qtp70666339-235] INFO org.apache.hadoop.hdds.utils.db.RDBCheckpointManage
r: Created checkpoint in rocksDB at /home/hadoop-001/ozonedata/ozonetmp1/db.checkpoints/om.db_chec
kpoint_1699502610434 in 27 milliseconds
2023-11-09 12:03:30,484 [qtp70666339-235] INFO org.apache.hadoop.hdds.utils.db.RDBCheckpointUtils:
 Waited for 21 milliseconds for checkpoint directory /home/hadoop-001/ozonedata/ozonetmp1/db.check
points/om.db_checkpoint_1699502610434 availability.
2023-11-09 12:03:30,484 [qtp70666339-235] INFO org.apache.hadoop.ozone.om.OMDBCheckpointServlet: C
ompaction pausing 1 ended. Elapsed ms: 65
2023-11-09 12:03:30,521 [qtp70666339-235] INFO org.apache.hadoop.hdds.utils.DBCheckpointServlet: Time taken to write the checkpoint to response output stream: 36 milliseconds
2023-11-09 12:03:30,521 [qtp70666339-235] INFO org.apache.hadoop.hdds.utils.DBCheckpointServlet: Excluded SST [] from the latest checkpoint.
2023-11-09 12:03:30,526 [qtp70666339-235] INFO org.apache.hadoop.hdds.utils.db.RocksDBCheckpoint: Cleaning up RocksDB checkpoint at /home/hadoop-001/ozonedata/ozonetmp1/db.checkpoints/om.db_checkpoint_1699502610434
2023-11-09 12:05:49,267 [qtp70666339-113] INFO org.apache.hadoop.hdds.utils.DBCheckpointServlet: Received GET request to obtain DB checkpoint snapshot
```

I will also submit the unit test code at the same time.

**Comment by aierate** [comment]
> Thanks for the patch @aierate
> 
> Was it an HA env where you were disabling ratis by setting `ozone.om.ratis.enable`? If I'm not wrong this code path is only reached in HA env and Ratis should be enabled in that scenario?

Thanks @hemantk-12 . My test environment is not an HA environment, and I only have one OM machine and three Datanode machines here. And this problem also occurred in our company's production environment.

**Comment by aierate** [comment]
Here is result of unit test with CI
https://github.com/aierate/ozone/actions/runs/6809417312
Thank for everyone review.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMDbCheckpointServlet.java [hunk 1]
```diff
@@ -285,6 +286,57 @@ public void testEndpoint(String httpMethod) throws Exception {
...
+    Assertions.assertTrue(
+        omMetrics.getDBCheckpointMetrics().
+            getLastCheckpointStreamingTimeTaken() > 0);
+    Assertions.assertTrue(omMetrics.getDBCheckpointMetrics().
+        getNumCheckpoints() > initialCheckpointCount);
+
+    Mockito.verify(omDbCheckpointServletMock).writeDbDataToStream(any(),
+        any(), any(), eq(toExcludeList), any(), any());
+  }
+
```
Comment (adoroszlai) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMDbCheckpointServlet.java:339]: Please do not duplicate the code from `testEndpoint()`.

```suggestion
  @ParameterizedTest
  @MethodSource("getHttpMethods")
  public void testEndpointNotRatis(String httpMethod) throws Exception {
    conf.setBoolean(OZONE_OM_RATIS_ENABLE_KEY, false);
    testEndpoint(httpMethod);
  }

```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMDBCheckpointServlet.java [hunk 2]
```diff
@@ -669,7 +669,9 @@ public BootstrapStateHandler.Lock lock()
       snapshotDeletingService.getBootstrapStateLock().lock();
 
       // Then wait for the double buffer to be flushed.
-      om.getOmRatisServer().getOmStateMachine().awaitDoubleBufferFlush();
+      if (om.isRatisEnabled()) {
+        om.getOmRatisServer().getOmStateMachine().awaitDoubleBufferFlush();
+      }
```
Comment (adoroszlai) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMDBCheckpointServlet.java:674]: `OzoneManager.awaitDoubleBufferFlush()` handles both cases (Ratis enabled/disabled).

```suggestion
      om.awaitDoubleBufferFlush();
```

**Review by adoroszlai — CHANGES_REQUESTED** [review]
Thanks @aierate for adding the test.  I would like to suggest some code improvements, but otherwise looks good.

**Comment by aierate** [comment]
This suggestion is very helpful to me, it is a very good suggestion, thank you very much, @adoroszlai.

**Comment by aierate** [comment]
> This suggestion is very helpful to me, it is a very good suggestion, thank you very much, @adoroszlai.

All the CI tests look passed. Could you help me merge the code? Thank you @adoroszlai .

**Review by adoroszlai — APPROVED** [review]

**Comment by adoroszlai** [comment]
Thanks @aierate for the contribution, @hemantk-12 for the review.

**Pull request merged**